### ISSUE TITLE
How to handle one DbContext across multiple repositories?

### ISSUE THREAD

**Issue by mrukas** [dotnet-architecture/eShopOnWeb#473]
I am using a similar approach than this in a project I am currently working on and experienced the following problem:

The repositories do share the same instance of the database context because usually the DbContext is getting registered as scoped (per request). The add, update and delete methods of the EfRepository class immediately save the changes by calling `SaveChangesAsync` on the context.

Couldn't this lead to confusing behaviour?
Lets say someone queries an entity (entity1) from a repository (repository1) . After that he makes a change to that entity (entity1) but does not immediately wants to save it. Now in the same request he wants to update another entity (entity2) from a different repository (repository).

The problem now is that the changes in entity1 automatically get saved, because the current database context is tracking the changes of entity1.
From a developers perspective it seems to be confusing because you have never called update on repository1 for entity1.

How do you tackle such a problem?

My solution would be to implement a unit of work pattern, that combines all repositories. Only the unit of work class would contain a `Save` method. I think it would be more clear for the developer, because someone would implicitly know, that changes made to entities, that got retrieved from the same unit of work would get saved all together.

In the current version i would consider the automatic save of entity1 a side effect. The developer would need detailed knowledge about the concrete implementation of the repository (Change Tracking in ef core).

Are there any other best practices, that could be used to solve that "problem"? I think disabling change tracking or registering the database context and the repositories as transient would be a bad solution.

Kind regards.

**Comment by netferret** [comment]
Usually scoped works well and sometimes I have used locking, best practice is to use the DB as little as possible and optimise, you can make it pretty fast, another approach is to offload to a queue then process each message.  Also it's quicker to use dapper and not EF which does lot of magic behind the scenes.  Use caching in memory and only persist when required.

Good luck.

**Comment by fiseni** [comment]
Hi @mrukas,

Yes, that's correct. On add/update/delete all changes in the DbContext will be saved. The Unit of Work will explicitly expose the Save method, but if you don't want to save the rest of the changes, than it won't fix the issue. On the other hand, you can add SaveChanges method explicitly in the repository as well (not to create additional UoW, since DbContext already doing the same job). And, be aware that no matter from which repository you're calling it, it still will save all the changes.

In any case, the devs always have to be aware of this. No matter if you're using UoW or not, if they're not aware what SaveChanges do under the hood, than there will always be improper usages.

If you actually want to exclude all the other changes, than you would need to flag the changed entities as not modified in the DbContext change tracker before save.

**Comment by ardalis** [comment]
Agreed with @fiseni . The reason for no explicit UoW implementation here is mainly for simplicity. Most web requests should be doing a very small thing and probably modifying at most 1 aggregate, so performing saves as part of the repository methods works well. In the edge cases where it does not (and they do exist of course), there are a number of options you can consider instead.

**Comment by mrukas** [comment]
@netferret 
Well usually EF works pretty great I think. I wouldn't want to change my whole ORM because of this problem. Dapper would also mean that I would need to write raw sql queries and therefore would be more bound towards a specific database engine than before. In my opinion a worker queue would be overkill to overcome this problem, especially if the software I am working is not really complex or huge.

@fiseni 
Yeah i know that DbContext is pretty much a UoW and DbSet pretty much are repositories. I just thought wrapping the repositories would make it more clear, that they are in fact one unit under the hood. I am also not really satisfied by using UoW, because it really would be overkill for simple applications and would add another layer. Therefore i wanted to know if there a different solutions or maybe best practices. I also thought about querying all entities with AsNotracking so the update methods would only mark the passed entity as modified. This would be pretty much the same as marking all other entities as "not modified".

@ardalis 
Thats totally true. For nearly all other web applications I was always dealing with one aggregate per request and therefore did not really experienced this problem. The software I am currently working however does use multiple repositories in the same DbContext scope and therefore this problem arose.

The thing that bothered me was that we now have clean interface for generic repositories, but someone would still need to know deep knowledge about the concrete implementation.

**Comment by escalonn** [comment]
The straightforward solution is to make the context's lifetime transient so that the repositories don't share the same one. Unless you are using a separate unit of work layer, using the same context across repositories _arguably_ breaks encapsulation, especially since the context isn't thread-safe so it's not expected that random changes would just appear in it from the perspective of a given thread.

**Comment by fiseni** [comment]
Querying with AsNoTracking for objects you actually intent to mutate, might be a bit harder than it seems.
You have to take care of owned types (if using any). Collections and related entities should be carefully handled as well. You might end up with scenario where on any small update, huge SQL update statements will be generated. Or, the scarier scenario, something to be ommited. Practically, you will end up implementing a change tracker.

I personally, won't suggest that option. Flaging everything rest as not modified is way easier than choosing carefully what to flag for modify (Collections can have add/update/delete, no matter the root is flagged as update).

**Comment by mrukas** [comment]
@escalonn 
I also thought about making it transient but I was not sure if this is really a great idea. I don't know how entity framework handles connections? Would that mean, that every DbContext instance uses a new database connection?

@fiseni 
I'm aware that this is not the easiest way. Calling update on the DbContext and passing the updated entity would probably be enough as long no referenced entity got deleted. I would have to take care that deleted entities would get marked as deleted. 
The problem I see when marking all other entites as "not modified" is that if I update one of these entities after the save i would also manually have to traverse the graph of these entities and manually mark the related entities as modified or deleted.

The scenario would be:

1. Repository1: Query entity1
2. Repository2: Query entity2
3. Changing properties on entity1 and entity2.
4. Call update on repository1 for entity1
   4.1 Mark all other entities as "not modified" also entity2.
   4.2 Save changes of entity1
5. Call update on repository2 for entity2.

Now only the changes of entity2 itself get saved if I'm using `_dbContext.Entry(entity).State = EntityState.Modified`. I could bypass this by using `context.Update(rootEntity);` which i mentioned above and is described [here](https://docs.microsoft.com/en-us/ef/core/saving/disconnected-entities#all-newall-existing-entities). Even than I would need to manually handle deletes.
So the two methods are essentially the same I think.
